src/charm.py

```diff
@@ -53,22 +63,143 @@ def __init__(self, *args):
-        self._create_resources()
-        self._patch_pod_labels()
         # Creates custom postgresql.conf file.
         self._patroni.render_postgresql_conf_file()
 
-    def _on_config_changed(self, _):
+    def _on_config_changed(self, _) -> None:
         """Handle the config-changed event."""
         # TODO: placeholder method to implement logic specific to configuration change.
         pass
 
-    def _on_leader_elected(self, _) -> None:
+    def _add_members(self, event) -> None:
+        """Add new cluster members.
+
+        This method is responsible for adding new members to the cluster
+        when new units are added to the application. This event is deferred if
+        one of the current units is copying data from the primary, to avoid
+        multiple units copying data at the same time, which can cause slow
+        transfer rates in these processes and overload the primary instance.
+        """
+        # Only the leader can reconfigure.
+        if not self.unit.is_leader():
+            return
+
+        # Reconfiguration can be successful only if the cluster is initialised
+        # (i.e. first unit has bootstrap the cluster).
+        if "cluster_initialised" not in self._peers.data[self.app]:
+            return
+
+        try:
+            # Compare set of Patroni cluster members and Juju hosts
+            # to avoid the unnecessary reconfiguration.
+            if self._patroni.cluster_members == self._hosts:
+                return
+
+            logger.info("Reconfiguring cluster")
+            self.unit.status = MaintenanceStatus("reconfiguring cluster")
+            for member in self._hosts - self._patroni.cluster_members:
+                logger.debug("Adding %s to cluster", member)
+                self.add_cluster_member(member)
+        except NotReadyError:
+            logger.info("Deferring reconfigure: another member doing sync right now")
+            event.defer()
+
+    def add_cluster_member(self, member: str) -> None:
+        """Add member to the cluster if all members are already up and running.
+
+        Raises:
+            NotReadyError if either the new member or the current members are not ready.
+        """
+        hostname = self._get_hostname_from_unit(member)
+
+        if not self._patroni.are_all_members_ready():
+            logger.info("not all members are ready")
+            raise NotReadyError("not all members are ready")
+
+        # Add the member to the list that should be updated in each other member.
+        self._add_to_endpoints(hostname)
+
+        # Add the labels needed for replication in this pod.
+        # This also enables the member as part of the cluster.
+        try:
+            self._patch_pod_labels(member)
+        except ApiError as e:
+            logger.error("failed to patch pod")
+            self.unit.status = BlockedStatus(f"failed to patch pod with error {e}")
+            return
+
+    @property
+    def _hosts(self) -> set:
```

Comment:
this property is missing a docstring

Reply:
Thanks. Fixed on https://github.com/canonical/postgresql-k8s-operator/pull/9/commits/249f4fda535bf1e54aa824196f0aaefed07a00e3.